snowfall/models/second_pass_model.py

```diff
@@ -0,0 +1,240 @@
+
+        # Note that len_per_seq does not count -1
+        len_per_seq = supervision_segments[:, 2].to(device)
+
+        if debug:
+            # offset indicates the arc start index of each seq
+            offset = k2.index(best_paths.arcs.row_splits(2),
+                              best_paths.arcs.row_splits(1))
+
+            # number of phones per seq.
+            # minus 1 to exclude the label -1
+            expected_len_per_seq = offset[1:] - offset[:-1] - 1
+
+            assert torch.all(torch.eq(len_per_seq, expected_len_per_seq))
+
+        # Note: `phones` also contains -1, for the arcs entering the final state
+        phones = best_paths.labels.clone()
+
+        # remove label -1
+        phones = phones[phones != -1]
+
+        # torch.split requires a tuple/list for sizes, so we use tolist() here.
+        phones_per_seq = torch.split(phones, len_per_seq.tolist())
+
+        # default padding value is 0
+        padded_phones = nn.utils.rnn.pad_sequence(phones_per_seq,
+                                                  batch_first=True)
+        # padded_phones is of shape (num_seqs, T)
+        # encoder_memory is of shape (T, num_batches, F)
+
+        # Number of frames T should be equal
+        assert padded_phones.shape[1] == encoder_memory.shape[0]
+        # Caution: number of seqs is not necessarily equal to number of batches
+        #  assert padded_phones.shape[0] == encoder_memory.shape[1]
+
+        encoder_memory = encoder_memory.permute(1, 0, 2)
+        # Now encoder_memory is (num_batches, T, F)
+
+        acoustic_out = []
+        for segment in supervision_segments.tolist():
+            batch_idx, start, duration = segment
+            end = start + duration
+            acoustic_tensor = encoder_memory[batch_idx, start:end]
+            acoustic_out.append(acoustic_tensor)
+
+        padded_acoustics = nn.utils.rnn.pad_sequence(acoustic_out,
+                                                     batch_first=True)
+        # padded_acoustics is of shape (num_seqs, T, F)
+
+        x2 = self.encoder_embed(padded_phones.long())
+        # x2 is (num_seqs, T, F)
+        x2 = self.encoder_pos(x2)
+
+        assert x2.shape == padded_acoustics.shape
+
+        # (B, T, F) -> (T, B, F)
+        x2 = x2.permute(1, 0, 2)
+
+        # compute two masks
+        # (1) padding_mask
+        # (2) src_mask for masked self-attention
+
+        key_padding_mask = _compute_padding_mask(len_per_seq)
+        # key_padding_mask is of shape (B, T)
+
+        attn_mask = generate_square_subsequent_mask(x2.shape[0]).to(device)
+        # attn_mask is of shape (T, T)
+
+        x2 = self.encoder(src=x2,
+                          mask=attn_mask,
+                          src_key_padding_mask=key_padding_mask)
+
+        # x2 is (T, B, F)
+
+        x2 = x2.permute(1, 0, 2)
+
+        # now x2 is (B, T, F)
+
+        x_concat = torch.cat((padded_acoustics, x2), dim=-1)
```

Comment:
TODO(fangjun): Use cross attention here, like decoders
- query: x2
- key and value: padded_acoustics